content/Stop creating services for everything - Knowing YAAF & the FormObject Pattern.md

```diff
@@ -0,0 +1,130 @@
+
+>-"Hey, I don't know how properly refactor this piece of code"
+>-"Dude, do a new service"
+
+**But no!** Making a new service is not always the best option, in some cases, we are _reinventing the wheel_ and maybe there is a pattern that already fits with our needs.
+
+So here comes **YAAF** (Yet Another Active Form) to save our day. **YAAF** is a gem that lets you create form objects in an easy and friendly Rails way. It makes use of `ActiveRecord` and `ActiveModel` features in order to provide you with a form object that behaves pretty much like a Rails model, and still be completely configurable.
+
+## When to use YAAF?
+
+Let's think that we have an API endpoint that saves a new post on our database. A post has a title, body, publisher, and could have tags and a category.
+
+Tags and categories could be created at the moment that the publisher sends the post. (If our inputs don't find the correct tag or category, they will let the user write the name of a new one).
+
+So in our controller, in the worst-case scenario we could have something like this:
+
+```ruby
+class Api::V1::PostsController < Api::V1::ApiController
+  def index
+    @posts = Post.all
+  end
+
+  def create
+    ActiveRecord::Base.transaction do
+      @post = Post.new(post_params)
+      @post.tags = params[:tags].map do |tag|
+        tag[:id].present? ? Tag.find(tag[:id]) : Tag.find_or_create_by(name: tag[:name])
+      end
+      if params[:category_name].present?
+        category = Category.create!(name: params[:category_name])
+        @post.category = category
+      end
+      @post.save!
+    end
+  end
+
+
+  def post_params
+    params.require(:post).permit(:title, :body, :publisher_id, :category_id)
+  end
+end
+```
+
+It looks terrible, doesn't it? Maybe our first thinking about that piece of code is to make a refactoring moving the creation of the post to a service named `PostCreationService`, which could be useful and it might be used in the future in another part of the system. But... what have we said about _reinventing the wheel_?
+
+Using **YAAF** we should create a new `PostForm` class that is going to encapsulate all the logic of post creation and related models inside it. And is very simple to implement it! Look at this code:
+```ruby
+# app/forms/registration_form.rb
+
+class PostForm < ApplicationForm
+  attr_accessor :post_form_params
+
+  def initialize(args = {})
+    @post_form_params = args
+    @models = [post, category, tags].flatten.compact
+  end
+
+  def post
+    @post ||= Post.new(post_form_params[:post])
+  end
+
+  def category
+    return [] if post_form_params[:category_name].blank?
+
+    post.category = Category.find_or_initialize_by(name: post_form_params[:category_name])
+    post.category
+  end
+
+  def tags
+    return [] if post_form_params[:tags].blank?
+
+    post.tags = post_form_params[:tags].map do |tag|
+      tag[:id].present? ? Tag.find(tag[:id]) : Tag.find_or_initialize_by(name: tag[:name])
+    end
+    post.tags
```

Comment:
```suggestion
    @tags ||= post_form_params[:tags].map do |tag|
      tag[:id].present? ? Tag.find(tag[:id]) : Tag.find_or_initialize_by(name: tag[:name])
    end
```

Note: This code change comes in combination with the code changes above (also test that it actually works 😂)